api/client/websocketupgradeproto/protocol.go

```diff
@@ -0,0 +1,344 @@
+					if err := c.writeFrame(
+						ws.NewCloseFrame(
+							ws.NewCloseFrameBody(ws.StatusNormalClosure, ""),
+						),
+					); err != nil {
+						if !isOkNetworkErrOrTimeout(err) {
+							c.logger.DebugContext(c.logContext, "error writing close frame", "error", err)
+						}
+					}
+				})
+			}
+			c.readError = io.EOF
+			return 0, io.EOF
+		case ws.OpBinary:
+			c.readBuffer = frame.Payload
+			return c.readLocked(b)
+		case ws.OpPong:
+			// Receives Pong as response to Ping. Nothing to do.
+		case ws.OpPing:
+			if c.connType == serverConnection {
+				continue
+			}
+			// If this is a client connection, we should respond with a Pong frame.
+			pongFrame := ws.NewPongFrame(frame.Payload)
+			if err := c.writeFrame(pongFrame); err != nil {
+				c.logger.DebugContext(c.logContext, "error writing Pong frame", "error", err)
+				return 0, err
+			}
+		}
+	}
+}
+
+// websocketCloseProtocol implements the WebSocket close protocol.
+// It sends a close frame to the client and waits for a close frame response.
+// This is important to ensure that the client receives the close frame and
+// can gracefully close the connection.
+// If the client does not respond, the connection will be closed after a deadline expires.
+// The acquireReadLock parameter indicates whether to acquire the read lock
+// before waiting for the close frame response. This is useful to prevent
+// concurrent reads while waiting for the close frame.
+func (c *WebsocketUpgradeConn) websocketCloseProtocol(closeCode ws.StatusCode, closeText string) {
+	if !c.supportsCloseProtocol {
+		return
+	}
+	c.once.Do(func() {
+		// Set a deadline to reset any previously set deadline.
+		// Also, this is important to ensure that the connection
+		// won't hang indefinitely if the client does not respond to the close frame.
+		const deadline = 3 * time.Second
+		if err := c.SetDeadline(time.Now().Add(deadline)); err != nil {
+			if !isOkNetworkErrOrTimeout(err) {
+				c.logger.DebugContext(c.logContext, "error setting read deadline", "error", err)
+			}
+			return
+		}
+		// Per RFC 6455, the side initiating the close should send a close frame
+		// and then wait for the close frame from the other side.
+		// If the other side does not respond, the connection will be closed after
+		// the deadline expires.
+		closeFrame := ws.NewCloseFrame(
+			ws.NewCloseFrameBody(closeCode, closeText),
+		)
+
+		if err := c.writeFrame(closeFrame); err != nil {
+			if !isOkNetworkErrOrTimeout(err) {
+				c.logger.DebugContext(c.logContext, "error writing close frame", "error", err)
+			}
+			return
+		}
+	})
+
+	c.readMutex.Lock()
+	defer c.readMutex.Unlock()
+	// Wait for the close frame from the other side.
+	// This will block until the close frame is received or the deadline expires.
+	// We will collect all data from the connection until we receive a close frame
+	// or an error occurs. This is important to ensure that we read all data
+	// from the connection before closing it.
+	var (
```

Comment:
any concern if there is already another live `Read` ?

Reply:
Read and this code share the same mutex. If any live Read is running and we do not receive the close confirmation we sent because the client is misbehaving, the set deadline above will cause the connection to time out after 3s